frame/clearing-house/src/tests/liquidate.rs

```diff
@@ -0,0 +1,449 @@
+use composable_traits::clearing_house::ClearingHouse;
+use frame_support::{assert_noop, assert_ok, traits::tokens::fungibles::Inspect};
+
+use super::{
+	as_balance, comp::approx_eq_lower, multi_market_and_trader_context,
+	traders_in_one_market_context, valid_market_config,
+};
+use crate::{
+	mock::{
+		accounts::{ALICE, BOB},
+		assets::USDC,
+		runtime::{
+			Assets as AssetsPallet, Oracle as OraclePallet, Origin, Runtime,
+			System as SystemPallet, TestPallet, Vamm as VammPallet,
+		},
+	},
+	tests::run_for_seconds,
+	Direction, Error, Event, FullLiquidationPenalty, FullLiquidationPenaltyLiquidatorShare,
+};
+
+// -------------------------------------------------------------------------------------------------
+//                                           Unit Tests
+// -------------------------------------------------------------------------------------------------
+
+#[test]
+fn cant_liquidate_user_with_no_open_positions() {
+	let config = valid_market_config();
+	let margins = vec![(ALICE, 0), (BOB, 0)];
+
+	// Set everything as usual, except that ALICE doesn't open any positions
+	traders_in_one_market_context(config, margins, |_| {
+		VammPallet::set_price(Some(1.into()));
+
+		// Bob can't liquidate Alice
+		assert_noop!(
+			TestPallet::liquidate(Origin::signed(BOB), ALICE),
+			Error::<Runtime>::UserHasNoPositions
+		);
+	});
+}
+
+#[test]
+fn cant_fully_liquidate_if_above_maintenance_margin_ratio_by_pnl() {
+	let mut config = valid_market_config();
+	config.margin_ratio_initial = (1, 2).into(); // 2x max leverage
+	config.margin_ratio_maintenance = (1, 10).into(); // 10% MMR
+	config.taker_fee = 0;
+
+	let margins = vec![(ALICE, as_balance(55)), (BOB, 0)];
+	traders_in_one_market_context(config, margins, |market_id| {
+		VammPallet::set_price(Some(100.into()));
+
+		// Alice opens a position
+		assert_ok!(
+			<TestPallet as ClearingHouse>::open_position(
+				&ALICE,
+				&market_id,
+				Direction::Long,
+				as_balance(100),
+				as_balance(1),
+			),
+			as_balance(1)
+		);
+
+		// Bob can't liquidate Alice
+		assert_noop!(
+			TestPallet::liquidate(Origin::signed(BOB), ALICE),
+			Error::<Runtime>::SufficientCollateral
+		);
+
+		// Price moves so that Alice's account is at exactly 10% margin ratio
+		// 100 -> 50
+		VammPallet::set_price(Some(50.into()));
+		// At price 50:
+		// - margin required = 5
+		// - PnL = -50
+		// - margin = 5
+
+		// Still not liquidatable; just enough collateral
+		assert_noop!(
+			TestPallet::liquidate(Origin::signed(BOB), ALICE),
+			Error::<Runtime>::SufficientCollateral
+		);
+	});
+}
+
+#[test]
+fn cant_fully_liquidate_if_above_maintenance_margin_ratio_by_funding() {
+	let mut config = valid_market_config();
```

Comment:
You could utilize a builder object instead of creating these mutable market_configs like you do. This wouldn't change the logic at all. I personally just like the look of 
let config = MarketConfigBuilder::default()
    .funding_frequency(60)
    .funding_period(60)
    ...
    .build();